app/src/features/authentication/public/hooks/useLoginForm.ts

```diff
@@ -13,69 +14,80 @@ export default function useLoginForm() {
+    const { toast } = useToast();
 
     const {
-        formState: { dirtyFields, errors, isDirty, isSubmitted, isValid },
-        handleSubmit,
+        formState: { dirtyFields, errors, isDirty, isSubmitted, isValid, isSubmitting, isSubmitSuccessful },
+        getValues,
         register,
+        trigger,
     } = useForm<LoginForm>({
         defaultValues: { password: '', username: '' },
     });
 
-    const errorMsg = useMemo(() => {
-        if (isSubmitted && !isValid) {
-            if (errors.password?.type && errors.username?.type) {
-                return 'Username and password are required';
-            }
-
-            if (errors.password?.type) {
-                return 'Password is required';
-            }
-
-            if (errors.username?.type) {
-                return 'Username is required';
-            }
-        }
-
-        if (isSubmitted && isValid && submitErrors) {
-            return 'Invalid username or password';
-        }
-    }, [isValid, isSubmitted, errors.password, errors.username, submitErrors]);
-
-    const hasError = !!errorMsg;
-
-    const onValidSubmit: SubmitHandler<LoginForm> = async (data: LoginForm, event) => {
+    const onValidSubmit = async (data: LoginForm, event: BaseSyntheticEvent<object, any, any>) => {
         const { password, username } = data;
 
         event?.preventDefault();
         setSubmitErrors(null);
-
         const result = await authenticate.signIn(username, password);
-
         if (isOk(result)) {
             navigate('/');
         } else {
             setSubmitErrors(result);
-            // forces the SubmitButton failure/shake state
-            throw result;
         }
+        return result;
     };
 
-    // TODO we need to "shake" the SubmitButton when the form is isSubmitted but not isValid
-    // adding an "invalid submit handler" as the 2nd arg here doesn't work, as throw'ing the
-    // errors to reject the promise causes the form state to get mucked up.
-    const onSubmit = handleSubmit(onValidSubmit);
+    const onInvalidSubmit = useCallback(
+        (event: BaseSyntheticEvent<object, any, any>) => {
+            event?.preventDefault();
+            let error = '';
+            if (errors.password?.type && errors.username?.type) {
+                error = 'Username and password are required';
+            } else if (errors.password?.type) {
+                error = 'Password is required';
+            } else if (errors.username?.type) {
+                error = 'Username is required';
+            }
+
+            if (isSubmitted && isValid && submitErrors) {
+                error = 'Invalid username or password';
+            }
+            console.log('error', error);
+            return error;
+        },
+        [errors, isSubmitted, isValid, submitErrors]
+    );
+
+    const onSubmit = useCallback(
+        async (event: BaseSyntheticEvent<object, any, any>) => {
+            trigger();
+
+            if (!isValid) {
```

Comment:
this is a bit curious that this works, since the `trigger()` to run validation wouldn't happen synchronously run such that the `isValid` state is already `false` in the same execution thread 🤔 

Does this work if you just immediately hit the submit button on page render w/o touching/dirtying the form inputs? I.e. im wondering if previous triggers of validation on input change/blur is whats putting the form in an invalid state.

Reply:
It does, though there are still issues with the right errors display. It seems to be a render behind on displaying the global errors